### DISCUSSION TITLE
Checked Exception으로 캡슐화가 깨지는 이유?

### DISCUSSION THREAD

**Discussion by CODe5753** [Well-Founded-Confidence/CleanCode#19] [미해결문제]
-

**Comment by sungyujeon** [comment]
개인적인 생각이지만..

캡슐화란 데이터나 로직 등에 접근하는 것을 숨기는 정보 은닉의 역할 뿐만이 아니라,
내부적인 데이터나 로직이 어떻게 돌아가는지는 관심이 없다는 것(숨겨져 있는 것은 신경을 쓰지 않아도 되는),
즉 캡슐화가 되어 있다면 호출이 되는 순간에는 특정 하나의 행위에만 책임이 있는 것이라고 생각이 됩니다.

예를 들어
array.sort()로 sort 메서드를 호출한다면
머지소트를 하던, 퀵소트를 하던, 팀소트를 하던 관심이 없고
array가 sorting 된다는 것에만 관심이 있다고 생각이 되는데요.

정렬을 하면서 여러가지 예외 상황이 발생할 수도 있고
이를 처리하는 코드들도 내부적으로 많을 수 있다고 봅니다.
그런데 만약에 checked exception이 발생한다면
array.sort()를 호출하는 메서드의 선언부에서도 명시적으로 해당 exception을 throws 해주는 등 예외 처리를 강제하게 될텐데요.

이렇게 된다면 array를 정렬한다는 관심사 외에 checked exception에 대한 예외 처리에도 신경을 써야 하기 때문에
캡슐화가 깨진다고 표현한 것이 아닌가 싶습니다.

이렇게 되면 궁금증이 생기는게 unchecked exception이 발생해도
예외 처리를 하는 것은 똑같은거 아니냐? 라는 의문이 생길 수도 있을 것 같습니다.
(내부적으로 잘못된 값을 넘겨서 IllegalArgumentException 등이 나는 상황?)

이러한 경우 위처럼 unchecked exception인 IllegalArgumentException이 발생하는 것은 runtime 시에 정상흐름을 벗어났을 때인데,
잘못된 input에 대해 잘못된 output(exception으로)을 던지는 일관된 흐름이지 않을까란 생각이 드네요.
(즉, 올바른 input -> 올바른 output, 잘못된 input -> 잘못된 output 이 보장되는)

그런데 checked exception의 경우
I/O 에러가 난다던가 StackOverflow가 난다던가 하는 것은
시스템적인 문제이기 때문에 어떨 때는 잘되다가 어떨 때는 잘못되다가 하는
일관성을 보장하지 못한다는 차이가 있는 것 같습니다.
그래서 예외를 강제하는게 아닌가란 생각도 드네요.

그런데 다시 돌아와서 생각해보면..
실제 코드에서 checked exception 던지는 메서드들은
그럼 모두 캡슐화가 깨진 것인가??? 하는 의문이 남네요..
